### PR TITLE
List all pods on node added for attach_detach controller

### PR DESCRIPTION
**Pull request by jsravn**
To ensure all pods with volumes are managed immediately upon a node
becoming managed. This most commonly happens at startup of the
controller manager.

This may produce duplicate pod processing, but this should be minimal
since nodes are generally added at slow rate. If it's a problem we'll
need to move to a worker queue approach and keep track of all nodes
so pods can be requeued for processing.

Main drawback is at startup. The attach detach controller may end up
processing all pods twice in the worse case. And without a node index
it will list all the pods for every node in the cluster.

### CODE REVIEW

**Comment by eparis** [comment]
@k8s-bot ok to test

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 1]
```diff
@@ -293,6 +294,23 @@ func (adc *attachDetachController) nodeAdd(obj interface{}) {
...
 	adc.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
+
+	if _, exists := node.Annotations[volumehelper.ControllerManagedAttachAnnotation]; exists {
+		// Newly added node is managed by controller, ensure all pods are up to date.
+		adc.processPodsForNode(nodeName)
+	}
+}
+
+func (adc *attachDetachController) processPodsForNode(nodeName types.NodeName) {
+	pods, err := adc.podLister.List(labels.Everything())
```
Comment (gnufied) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: Is there a way we can use a selector to pick only pods which belong to this node? Not a blocker, but this defenitely looks promising.

**Comment by cblecker** [comment]
```
I0419 09:31:36.636] Verifying hack/make-rules/../../hack/verify-bazel.sh
I0419 09:31:41.586] 33a34
I0419 09:31:41.587] >         "//vendor/k8s.io/apimachinery/pkg/labels:go_default_library",
W0419 09:31:41.687] DRY-RUN: wrote "pkg/controller/volume/attachdetach/BUILD"
W0419 09:31:42.228] 
W0419 09:31:42.228] 1 BUILD files not up-to-date.
I0419 09:31:42.329] 
I0419 09:31:42.329] Run ./hack/update-bazel.sh
I0419 09:31:42.329] FAILED   hack/make-rules/../../hack/verify-bazel.sh	6s
```

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 1 above)
Comment (kargakis) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: We could index the pod cache based on a NodeNameIndex and fetch pods for a specific node but I haven't looked close wrt the impact of using multiple indexes in cache. https://github.com/kubernetes/kubernetes/issues/38868

Comment (jsravn) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: This is a problem at startup especially. For every node in the cluster it will do a list of the pods. I notice we do lots of list alls though (like in daemonset controller), so I'm not sure if it's a big deal or not.

**Comment by jsravn** [comment]
How do I get the bazel build to find the added import?

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 2]
```diff
@@ -293,6 +294,24 @@ func (adc *attachDetachController) nodeAdd(obj interface{}) {
 	// the attached volumes field. This function ensures that we sync with
 	// the actual status.
 	adc.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
+
+	if _, exists := node.Annotations[volumehelper.ControllerManagedAttachAnnotation]; exists {
```
Comment (jsafrane) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: Is ADC locked at this time? I am not sure, but I think the reconciler runs in a separate goroutine and it could run at this point, where we have node and its attached volumes in states of the world, while the pods are not there yet. reconcile() will try detach these volumes.

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 1 above)
Comment (jsafrane) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: podLister works with shared informer cache, it won't ask API server. So it's IMO OK for now and it should be fixed in a subsequent PR.

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 2 above)
Comment (jsravn) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: I think you're right. That race already exists though, correct? At least by immediately adding the pod information here, we have a shot at avoiding it. If we do hit the race, worse case is it will detach then re-attach when the pods are processed.

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 1 above)
Comment (jsravn) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: I'm happy to work on that, I agree though it should be a separate PR, esp since I don't think it can be easily cherry picked into older releases.

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 2 above)
Comment (jsravn) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: Actually I think waitForCacheSync in #run will help here. It will prevent the reconciler from running until all the nodes/pods have been processed. Still a small race since the handlers operate asynchronously so may still be running even when waitForCacheSync returns.

**Comment by jsravn** [comment]
I pinged @saad-ali to have a look

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 3]
```diff
@@ -250,10 +255,6 @@ func (adc *attachDetachController) podAdd(obj interface{}) {
 	if pod == nil || !ok {
 		return
 	}
-	if pod.Spec.NodeName == "" {
```
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: Why remove this?

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 4]
```diff
@@ -293,6 +294,24 @@ func (adc *attachDetachController) nodeAdd(obj interface{}) {
...
 	adc.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
+
+	if _, exists := node.Annotations[volumehelper.ControllerManagedAttachAnnotation]; exists {
+		// Newly added node is managed by controller, ensure all pods are up to date.
+		// Run in a separate goroutine so we don't block the node handler goroutine.
+		go adc.processPodsForNode(nodeName)
+	}
+}
+
+func (adc *attachDetachController) processPodsForNode(nodeName types.NodeName) {
```
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: rename to `fetchAndProcessPodsForNode(...)` -- I want the name to emphasis that it is not passively processing but actively fetching.

Also add a comment summarizing what this method does above it.

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 1 above)
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: This is adding a lot of work to the controller: if there are 5000 nodes with 30 pods each, when the controller restarts, each node you will fetch and process 150,000 pods, meaning it will end up processing 450,000 pods (excluding the ones it will get from watch).

Given that we have https://github.com/kubernetes/kubernetes/pull/42033, which periodically lists all pods, do we really needs this as well to fix https://github.com/kubernetes/kubernetes/issues/43300?

**Review by saad-ali — COMMENTED** [review]
My big question is about if we should do this at all

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 3 above)
Comment (jsravn) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: It's redundant with ProcessPodVolumes, which does the same check (with a log message).

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 1 above)
Comment (jsravn) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: Yeah we'd need a node index to do this efficiently, or use a work queue w/ pod processing retries. Considering #42033 also processes every single pod every 3 mins is it a big deal in comparison? I wasn't sure.

The main purpose for me is to get a fix we can use in 1.6/1.5. I didn't think #42033 was a candidate for backporting, or was even going to stick around in its current form. I thought it was more of a backup, experimental solution. Is it okay to have to wait for the reconciler to fix things when new nodes get added?

If we just want #42033, that's okay with me, as long as we can get a fix in 1.6/1.5 as well, since this is a major problem in those releases.

**Comment by saad-ali** [comment]
> Yeah we'd need a node index to do this efficiently, or use a work queue w/ pod processing retries. Considering #42033 also processes every single pod every 3 mins is it a big deal in comparison? I wasn't sure.
> 
> The main purpose for me is to get a fix we can use in 1.6/1.5. I didn't think #42033 was a candidate for backporting, or was even going to stick around in its current form. I thought it was more of a backup, experimental solution. Is it okay to have to wait for the reconciler to fix things when new nodes get added?
> 
> If we just want #42033, that's okay with me, as long as we can get a fix in 1.6/1.5 as well, since this is a major problem in those releases.

The reason I prefer PR #42033 over this PR is:
1) #42033 seems more robust--it will continuously ensure state is not skewed over the lifetime of the controller process instead of just at startup,
2) #42033 seems like a smaller (although still non-trivial) performance hit: it processes potentially 5000 nodes * 30 pods =150,000 pods at 3 minute increments vs this solution processes potentially 5000 nodes * 30 pods = 150,000 pods * 5000 nodes = 750,000,000 pods all up front--that's a three order of magnitude difference processed all up front.

I understand that PR #42033 may be more difficult to port back to older branches, but because of the reasons above and that PR #42033 appears to address the problem raised in issue #43300 (making attach/detach controller tolerant to HA failover), I hesitate to endorse this PR.

**Comment by jsravn** [comment]
@saad-ali 
> #42033 seems more robust--it will continuously ensure state is not skewed over the lifetime of the controller process instead of just at startup,

This also takes care of processing pods when node state changes. Like adding nodes to the cluster.

> #42033 seems like a smaller (although still non-trivial) performance hit: it processes potentially 5000 nodes * 30 pods =150,000 pods at 3 minute increments vs this solution processes potentially 5000 nodes * 30 pods = 150,000 pods * 5000 nodes = 750,000,000 pods all up front--that's a three order of magnitude difference processed all up front.

Agreed, so I think we'd need the node indexing. @jsafrane suggested we do that in a separate PR.

So do you think we should just try backporting #42033? Is it something @NickrenREN could pick up?

**Comment by saad-ali** [comment]
> This also takes care of processing pods when node state changes. Like adding nodes to the cluster.

I think #42033 would handle that case too: add goes from unmanaged to attach/detach controller managed, controller gets an add, adds pod to desired state. Populator comes around, find that pods that were previously being ignored now have a node in desired state, adds them.

> Agreed, so I think we'd need the node indexing. @jsafrane suggested we do that in a separate PR.

Could you expand on what you mean by node indexing?

> So do you think we should just try backporting #42033? Is it something @NickrenREN could pick up?

Yes, let's backport it but let's ensure that #42033 hasn't caused any major regressions. @NickrenREN is this something you can help with?

**Comment by jingxu97** [comment]
For a cluster with M nodes and N pods per node, when controller restarted, controller will process M * ( M*N) pods during startup. Will it cause high overhead?

**Comment by jsravn** [comment]
@saad-ali 
> Could you expand on what you mean by node indexing?

I mean adding a method to PodLister so we can list by nodename, like we can do for namespace. It adds an Indexer for node on pods. There was another use case for this as well, I think in the daemonset controller.

**Comment by NickrenREN** [comment]
@saad-ali  I'd like to help if needed

**Comment by jsravn** [comment]
We can close this then if we're deciding to backport #42033.

**Comment by gnufied** [comment]
I have opened a cherry-pick of #42033 - https://github.com/kubernetes/kubernetes/pull/45620

@NickrenREN can you please review?

**Comment by jsravn** [comment]
Closing in favour of #45620
